### PR TITLE
binutils: 2.31.1 -> 2.32

### LINKED ISSUE #51293: gdb does not work on macOS Mojave
## Issue description

`gdb` does not work on on macOS Mojave, [because Apple has introduced a new load command to to o-mach executables.](https://stackoverflow.com/questions/52529838/gdb-8-2-cant-recognized-executable-file-on-macos-mojave-10-14). A [patch](https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;h=fc7b364aba41819a5d74ae0ac69f050af282d057) has been applied upstream to fix this problem.

### Steps to reproduce

Run `gdb` on any executable file.

```
% gdb /usr/bin/true 
GNU gdb (GDB) 8.1.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-apple-darwin".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
BFD: /usr/bin/true: unknown load command 0x32
BFD: /usr/bin/true: unknown load command 0x32
BFD: /usr/bin/true: unknown load command 0x32
"/usr/bin/true": not in executable format: Format de fichier non reconnu
```
## Technical details

 - system: `"x86_64-darwin"`
 - host os: `Darwin 18.2.0, macOS 10.14.1`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.1.3`
 - channels(smaret): `"nixpkgs-19.03pre160749.3a393eecafb"`
 - nixpkgs: `/Users/smaret/.nix-defexpr/channels/nixpkgs`

### PR DESCRIPTION
**Pull request by matthewbauer**
Fixes #51293.

### CODE REVIEW

**Comment by matthewbauer** [comment]
I am getting some unexplained errors when using this. Probably better to figure those out before merging.

**Comment by dtzWill** [comment]
cc #55165 FWIW, likely same but haven't checked :).

File: pkgs/development/tools/misc/binutils/default.nix [hunk 1]
```diff
@@ -9,10 +9,7 @@
 let
   reuseLibs = enableShared && withAllTargets;
 
-  # Remove gold-symbol-visibility patch when updating, the proper fix
-  # is now upstream.
-  # https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=commitdiff;h=330b90b5ffbbc20c5de6ae6c7f60c40fab2e7a4f;hp=99181ccac0fc7d82e7dabb05dc7466e91f1645d3
-  version = "2.31.1";
```
Comment (dtzWill) [inline pkgs/development/tools/misc/binutils/default.nix]: may wish to revise commit message since this starts with 2.31.1 not 2.30.

**Comment by dtzWill** [comment]
What errors? :grin:

Not entirely unexpected, but if you get to a point where details make
sense (or perhaps hydra links) please share! :)

**Comment by matthewbauer** [comment]
Let me find the logs…

**Comment by matthewbauer** [comment]
@dtzWill here is the error I get building coreutils on x86_64-linux with this change:

```
mkdir: cannot change permissions of 'parent/sub': Function not implemented
parent/sub: expected d-wx--x-w-, got d-wx----w-
mkdir: cannot change permissions of 'parent/sub': Function not implemented
parent/sub: expected d-wx--x-w-, got d-wx----w-
FAIL tests/mkdir/perm.sh (exit status: 1)
```

It could be the machine though. I'd be interested in seeing if anyone else has the same issue.

**Comment by vcunat** [comment]
Have you considered using [later 2.31 branch](https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=shortlog;h=refs/heads/binutils-2_31-branch)?  (no tarballs unfortunately)  It happened multiple times for us that fresh binutils caused lots of failures.

**Comment by dtzWill** [comment]
That's likely more appropriate, in terms of getting fixes somewhat
safely.

I'd expect when/if we move to 2.32 we'll really be moving to
2.32+patches or hopefully the 2.32 branch :).

Agreed moving to new versions often introduces problems,
must be done carefully with a need for more thorough testing
than pretty much any other package :).

Sounds like we should perhaps start looking at what things look
like using the branch for 2.31?

(I haven't looked closely, but until there's a feature we need
from 2.32 and 2.31 is getting fixes... that seems the way to go)

**Pull request closed without merging**